src/Peachpie.Library/Hash.cs

```diff
@@ -2618,16 +2622,265 @@ public static PhpString hash_pbkdf2(string algo, byte[] password, byte[] salt, i
+        {
+            switch (algo)
+            {
+                // Default
+                case 0:
+                // Blowfish
+                case 1:
+                    return HashPasswordBlowfish(password, opt);
+                // Argon2i
+                case 2:
+                case 3:
+                    return HashPasswordArgon2(password, opt, algo == 2);
+                // Unknown algorithm
+                default:
+                    return PhpValue.False;
+            }
+        }
 
         /// <summary>
         /// Verifies that a password matches a hash.
         /// </summary>
         public static bool password_verify(string password, string hash)
         {
+            try
+            {
+                if (hash.Substring(1, 7) == "argon2i")
+                    return Argon2.Verify(hash, password);
+            }
+            catch (Exception)
+            {
+                return false;
+            }
             return crypt(password, hash) == hash;
         }
 
+        /// <summary>
+        /// This function checks to see if the supplied hash implements the algorithm and options provided. If not, it is assumed that the hash needs to be rehashed.
+        /// </summary>
+        /// <param name="hash">A hash created by password_hash()</param>
+        /// <param name="algo">A password algorithm constant denoting the algorithm to use when hashing the password. 0 - Default, 1 - Blowfish, 2 - Argon2i, 3 - Argon2id</param>
+        /// <param name="opt">See the password algorithm constants. If omitted, a random salt will be created and the default cost will be used.</param>
+        /// <returns>Returns TRUE if the hash should be rehashed to match the given algo and options, or FALSE otherwise.</returns>
+        public static bool password_needs_rehash(string hash, int algo, PhpArray opt = null)
+        {
+            bool result = false;
+
+            switch (algo)
+            {
+                // Default
+                case 0:
+                // BCrypt
+                case 1:
+                    string[] hashParts = hash.Split('$');
+
+                    if (hashParts.Length >= 3 && hashParts[1][0] == '2' && hashParts[1].Length >= 2) // $2 $ Right algorithm
+                    {
+
+                        if (opt != null && opt.ContainsKey("cost")) // Check options
+                        {
+                            PhpValue pom = opt.GetItemValue(new IntStringKey("cost"));
+                            result = !(hashParts[2] == pom.ToString());
+                        }
+
+                    }
+                    else
+                        result = true;
+                    break;
+                // Argon2i
+                case 2:
+                // Argon2id
+                case 3:
+                    Regex reg = new Regex(@"^\$(argon2id|argon2id)\$v=\d+\$m=(\d+),t=(\d+),p=(\d+)\$.+\$.+$");
```

Comment:
Create a static field from ```reg``` so that it doesn't have to be recreated with each call to the function.